src/qibo/models/utils.py

```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
+import numpy as np
+
+from qibo import gates as maingates
+from qibo.gates import gates
+from qibo.models.circuit import Circuit
+
+
+def convert_bit_to_energy(hamiltonian, bitstring):
+    """
+    Given a binary string and a hamiltonian, we compute the corresponding energy.
+    """
+    n = len(bitstring)
+    c = Circuit(n)
+    for i in range(n):
+        c.add(gates.X(int(i)))
+    result = c()  # this is an execution result, a quantum state
+    return hamiltonian.expectation(result.state())
+
+
+def convert_state_to_count(state):
+    """
+    This is a function that convert a quantum state to a dictionary keeping track of
+    energy and its frequency.
+    d[energy] records the frequency
+    """
+    m = int(np.log2(state.size))
+    c = Circuit(m)
+    c.add(maingates.M(*[i for i in range(m)]))
+    result = c(state, nshots=100)
+    counts = result.frequencies(binary=True)
+    return counts
+
+
+def compute_cvar(probabilities, values, alpha):
+    """
+    Auxilliary method to computes CVaR for given probabilities, values, and confidence level.
+
+    Args:
+        probabilities (list): list/array of probabilities
+        values (list): list/array of corresponding values
+        alpha (float): confidence level
+
+    Returns:
+        CVaR
+    """
+    sorted_indices = np.argsort(values)
+    probs = np.array(probabilities)[sorted_indices]
+    vals = np.array(values)[sorted_indices]
+    cvar = 0
+    total_prob = 0
+    for i, (p, v) in enumerate(zip(probs, vals)):
+        done = False
+        if p >= alpha - total_prob:
+            p = alpha - total_prob
+            done = True
+        total_prob += p
+        cvar += p * v
+    cvar /= total_prob
+    return cvar
+
+
+def cvar(hamiltonian, state, alpha=0.1):
+    """
+    Given the hamiltonian and state, this function estimate the
+    corresponding cvar function
+    """
+    counts = convert_state_to_count(state)
```

Comment:
Here you are calculating the probabilities via a non-deterministic measurement (like emulating quantum hardware). Is this on purpose? Because you could also extract the probabilities deterministically from the state vector without the need of measurements.

Reply:
thanks. let me change the code to make it deterministically.

Reply:
Hi Stavros, it is easy to compute the energy with the corresponding probabilities in Qibo?

Reply:
If I understand correctly there are two parts of the calculation:
1. `convert_state_to_count`: takes a statevector and returns frequencies. Then these frequencies are used to calculate probabilities. You could avoid this step by calculating probabilities directly as `np.abs(state) ** 2`. However these probabilities will be exact (as if nshots -> infinity), while in your calculation there will be some statistical error due to using finite nshots. I am not sure if you need this error in your calculation for some reason.
2. `convert_bit_to_energy`: calculates the energy of a product state: <bitstring|H|bitstring>. By definition this is a diagonal element in the matrix representation of the Hamiltonian in the computational basis. So if you have the matrix, it is probably easier and faster if you just convert the bitstring to decimal and index the matrix diagonal: `hamiltonian.matrix[decimal, decimal]`. If you don't have the matrix (eg. you are using `SymbolicHamiltonian`) what you are doing may be faster.

Reply:
I changed my `convert_state_to_count` code to compute the probability exactly. Having some macos 3.10 installation issues. Currently, I am trying to figure out

Reply:
> Having some macos 3.10 installation issues. Currently, I am trying to figure out

Don't worry about this, it is not related to this PR. There are some issues with dill tests and mac which are associated with new releases of some dependencies and we will fix them in another PR.

Reply:
Is this pull request alright? Anymore changes that I should make?

Reply:
@shangtai I believe so, yes. Could you please fix the conflicts? I believe they are due to some pre-commit update.

Reply:
@stavros11  Thanks. I have fixed the conflicts.